apps/mymedicare_cb/models.py

```diff
@@ -1,76 +1,106 @@
 import logging
-from django.db import models
+from django.db import models, transaction
 from django.contrib.auth.models import User, Group
 from rest_framework import exceptions
 from apps.accounts.models import UserProfile
-from apps.fhir.authentication import convert_sls_uuid
-from apps.fhir.server.authentication import authenticate_crosswalk
-from apps.fhir.bluebutton.exceptions import UpstreamServerException
-from apps.fhir.bluebutton.models import Crosswalk
+from apps.fhir.server.authentication import match_hicn_hash
+from apps.fhir.bluebutton.models import Crosswalk, hash_hicn
 from apps.fhir.bluebutton.utils import get_resourcerouter
 
 logger = logging.getLogger('hhs_server.%s' % __name__)
 
 
 def get_and_update_user(user_info):
-    username = convert_sls_uuid(user_info['sub'])
+    """
+    Find or create the user associated
+    with the identity information from the ID provider.
+
+    Args:
+        user_info: Identity response from the userinfo endpoint of the ID provider.
+
+    Returns:
+        A User
+
+    Raises:
+        KeyError: If an expected key is missing from user_info.
+        KeyError: If response from fhir server is malformed.
+        AssertionError: If a user is matched but not all identifiers match.
+    """
+    subject = user_info['sub']
+    hicn = user_info['hicn']
+    hicn_hash = hash_hicn(hicn)
```

Comment:
Suggest that we define a validation test for properly formatted HICNs. We use this test here and elsewhere in the code.

Reply:
agreed: https://jira.cms.gov/browse/BLUEBUTTON-1695